src/tree/updater_gpu_hist.cu

```diff
@@ -1326,23 +1305,50 @@ inline void DeviceShard<GradientSumT>::InitCompressedData(
 
 template <typename GradientSumT>
 inline void DeviceShard<GradientSumT>::CreateHistIndices(
-    const SparsePage& row_batch, size_t row_stride, int null_gidx_value) {
+    const SparsePage &row_batch,
+    const common::HistCutMatrix &hmat,
+    const std::vector<size_t> &shard_allocations,
+    int rows_per_batch) {
+  // Has any been allocated for me in this batch?
+  size_t num_elems = shard_allocations[shard_idx];
+  if (!num_elems) return;
+
+  size_t row_stride = this->ellpack_matrix.row_stride;
+
+  // Take into account the rows that could be processed by other GPUs
+  size_t begin_idx(0);
+  for (size_t i = 0; i < shard_idx; ++i) begin_idx += shard_allocations[i];
+
+  const auto& offset_vec = row_batch.offset.ConstHostVector();
+  /*! \brief row offset in SparsePage (the input data). */
+  thrust::device_vector<size_t> row_ptrs(num_elems+1);
+  thrust::copy(offset_vec.data() + begin_idx,
+               offset_vec.data() + begin_idx + num_elems + 1,
+               row_ptrs.begin());
+
   int num_symbols = n_bins + 1;
   // bin and compress entries in batches of rows
-  size_t gpu_batch_nrows =
+  size_t gpu_batch_nrows = 0;
+  if (rows_per_batch == 0) {
+     gpu_batch_nrows =
       std::min
       (dh::TotalMemory(device_id) / (16 * row_stride * sizeof(Entry)),
-       static_cast<size_t>(n_rows));
-  const std::vector<Entry>& data_vec = row_batch.data.HostVector();
+       static_cast<size_t>(num_elems));
+  } else if (rows_per_batch == -1) {
+     gpu_batch_nrows = num_elems;
+  } else {
+     gpu_batch_nrows = std::min(num_elems, static_cast<size_t>(rows_per_batch));
+  }
+  const std::vector<Entry>& data_vec = row_batch.data.ConstHostVector();
 
   thrust::device_vector<Entry> entries_d(gpu_batch_nrows * row_stride);
-  size_t gpu_nbatches = dh::DivRoundUp(n_rows, gpu_batch_nrows);
+  size_t gpu_nbatches = dh::DivRoundUp(num_elems, gpu_batch_nrows);
```

Comment:
What does `num_elems` refer to here? It looks like it is referring to some number of rows but the word element normally refers to a single entry in a matrix not a row.

Reply:
reworded.